### PR TITLE
Skip Upper vector save/restore for blocks that THROW

### LINKED ISSUE #65332: Assertion failed 'block->KindIs(BBJ_NONE, BBJ_ALWAYS)'
Failures in superpmi-replay with `JitStressRegs=1` and `JitStressRegs=8`. This is causing the pipeline to run "red": https://dev.azure.com/dnceng/public/_build?definitionId=1055

Example [log](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-d799c9c3cf774fe29b/linux-arm64/1/console.c0629dc0.log?sv=2019-07-07&se=2022-03-06T10%3A57%3A30Z&sr=c&sp=rl&sig=hs9GleKy%2BjBDJ7vTp9%2F5lGDv9SHpJy2XTRLWGgjibGM%3D):
```
ISSUE: <ASSERT> #66689 D:\a\_work\1\s\src\coreclr\jit\lsra.cpp (6435) - Assertion failed 'block->KindIs(BBJ_NONE, BBJ_ALWAYS)' in 'Roslyn.Utilities.SetWithInsertionOrder`1[Vector`1][System.Numerics.Vector`1[System.Single]]:Insert(int,System.Numerics.Vector`1[Single]):bool:this' during 'LSRA allocate' (IL size 78)
```

Looks like it fails for arm64 and arm32 replays.

### PR DESCRIPTION
**Pull request by kunalspathak**
Currently, we save and restore vectors for `BBJ_THROW` blocks as well, which is not really needed. In https://github.com/dotnet/runtime/pull/62662, we avoiding doing save/restore for calls that do not return. This PR extends the logic to also skip save/restore if the block ends with throw.

This also reverts https://github.com/dotnet/runtime/pull/66062.

Fixes: https://github.com/dotnet/runtime/issues/65332

### CODE REVIEW

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib , @BruceForstall

File: src/coreclr/jit/lsrabuild.cpp [hunk 1]
```diff
@@ -1456,6 +1456,34 @@ Interval* LinearScan::getUpperVectorInterval(unsigned varIndex)
...
+// Arguments:
+//    tree       - The current node being handled
+//
+// Notes: Vector save/restore is usually not needed before the call that never returns.
+//        It is also not needed if the block ends with a throw.
+bool LinearScan::shouldBuildUpperVectorSaveRestore(GenTree* tree)
+{
+    if ((tree != nullptr) && tree->IsCall())
+    {
+        if (tree->AsCall()->IsNoReturn())
```
Comment (EgorBo) [inline src/coreclr/jit/lsrabuild.cpp:1471]: You might also want to check `fgIsThrow` here but I assume both `IsNoreReturn` and `fgIsThrow` should always be inside a `BBJ_THROW` block

File: src/coreclr/jit/lsrabuild.cpp [hunk 2]
```diff
@@ -1456,6 +1456,34 @@ Interval* LinearScan::getUpperVectorInterval(unsigned varIndex)
...
+        {
+            // No point in having vector save/restore if the call will not return.
+            return false;
+        }
+    }
+
+    if ((compiler->compCurBB != nullptr) && (compiler->compCurBB->KindIs(BBJ_THROW)))
```
Comment (EgorBo) [inline src/coreclr/jit/lsrabuild.cpp:1478]: Perhaps, block should a parameter, or method renamed to something stating it checks current block?

File: src/coreclr/jit/lsrabuild.cpp (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsrabuild.cpp:1471]: I looked around and seems that the only non-throw case where calls are marked as NoReturn is here:
https://github.com/dotnet/runtime/blob/24ce527fb5f966a8262e48b5812cefc240ba6ef5/src/coreclr/jit/fgbasic.cpp#L2138

I am not familiar with how that translates to during LSRA.

File: src/coreclr/jit/lsrabuild.cpp (hunk 2 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsrabuild.cpp:1478]: I thought about passing block as a parameter but there are lot of call chain where I would have to pass `block`. We use `compCurBB` in earlier phases as well as `curBBNum` in LSRA to avoid doing that. As far method name, I think I have stated in the comments that it inspects the block.

File: src/coreclr/jit/lsrabuild.cpp [hunk 3]
```diff
@@ -1456,6 +1456,34 @@ Interval* LinearScan::getUpperVectorInterval(unsigned varIndex)
...
+//
+// Notes: Vector save/restore is usually not needed before the call that never returns.
+//        It is also not needed if the block ends with a throw.
+bool LinearScan::shouldBuildUpperVectorSaveRestore(GenTree* tree)
+{
+    if ((tree != nullptr) && tree->IsCall())
+    {
+        if (tree->AsCall()->IsNoReturn())
+        {
+            // No point in having vector save/restore if the call will not return.
```
Comment (BruceForstall) [inline src/coreclr/jit/lsrabuild.cpp:1473]: It's not clear to me why this would be safe. Couldn't you have computation after a call that requires restoring the registers, and before a throw?

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/lsrabuild.cpp (hunk 3 above)
Comment (BruceForstall) [reply src/coreclr/jit/lsrabuild.cpp:1473]: Seems like it's only safe if there are not uses of the (restored) register between this point and the throw, in which case the restore is "dead code".

Comment (kunalspathak) [reply src/coreclr/jit/lsrabuild.cpp:1473]: Vector restore happens at two places: 
- At the end of every block, whenever any `largeVectorVars` was partially spilled
- At the locations where we see the use

While it should be safe to eliminate the restores (and hence the saves) that happen for #1 above if the block is a THROW block, but there can be a scenario (as you pointed out) where eliminating save/restore will be wrong:

```asm
save(Vx)  ; <- shouldn't be eliminated because of the use
CALL ...
Vx = restore() ;  <- shouldn't be eliminated because of the use
 = use(Vx)
...
THROW
```

Since UpperVectorSave refpositions are created when we build kills, it is hard to look ahead to see which ones of them will be used (and hence restored) and thus can be eliminated. For e.g. in following case, we can (should) eliminate the unnecessary save/restore.

```asm
save(Vx)  ; <- safe to eliminate
CALL ...
Vx = restore() ;  <- safe to eliminate
...
THROW
```

To summarize, I was trying to eliminate vector saves, but it might not be safe (as Bruce pointed), so I will just try to at least eliminate the restore that happens at the end of the THROW block.

Comment (kunalspathak) [reply src/coreclr/jit/lsrabuild.cpp:1473]: Hhm, looks like our model is not built such that it can withstand to see only saves and not (optimally removed) restores.

Comment (kunalspathak) [reply src/coreclr/jit/lsrabuild.cpp]: Pushed an attempt to track for a block that throws, if a spill done before the call will ever be used. If not, skip save/restore for them. While I am at it, I will check what it takes to fix the problem @sandreenko pointed out in https://github.com/dotnet/runtime/pull/59315.

**Comment by kunalspathak** [comment]
The failures seems to be infrastructure issue. superpmi-replay is green which used to fail earlier. 
@BruceForstall - can you take a look at the fresh updates?

**Review by BruceForstall — APPROVED** [review]
LGTM

**Comment by kunalspathak** [comment]
Failure is https://github.com/dotnet/runtime/issues/65455.
